address snpEff database build failures

Sometimes the snpEff test can fail:

```
test/integration/test_intrahost.py::TestSnpEff::test_snpeff 00:00:00    ERROR while connecting to http://downloads.sourceforge.net/pro\
ject/snpeff/databases/v4_3/snpEff_v4_3_35420824d4be308c07afde83d507288ed3f5395596628c14fb8f38e.zip
java.lang.RuntimeException: java.lang.RuntimeException: File not found on the server. Make sure the database name is correct.
       at org.snpeff.util.Download.download(Download.java:178)
       at org.snpeff.snpEffect.commandLine.SnpEffCmdDownload.downloadAndInstall(SnpEffCmdDownload.java:32)
       at org.snpeff.snpEffect.commandLine.SnpEffCmdDownload.runDownloadGenome(SnpEffCmdDownload.java:86)
       at org.snpeff.snpEffect.commandLine.SnpEffCmdDownload.run(SnpEffCmdDownload.java:72)
       at org.snpeff.SnpEff.run(SnpEff.java:1183)
       at org.snpeff.SnpEff.main(SnpEff.java:162)
Caused by: java.lang.RuntimeException: File not found on the server. Make sure the database name is correct.
       at org.snpeff.util.Download.download(Download.java:127)
       ... 5 more

[gw1] [100%] FAILED test/integration/test_intrahost.py::TestSnpEff::test_snpeff
```

This happens when there are database hash lines in the snpEff config (`<conda_env>/share/snpeff-4.3.1t-1/snpEff.config`, toward the bottom), but no database directory named with same hash within `<conda_env>/share/snpeff-4.3.1t-1/data/`. snpEff checks for databases within `data/` first based on database names within the config file, and then attempts to download them if it can't find a corresponding database directory. It downloads pre-built databases from the snpEff sourceforge site rather than GenBank, which is why we go through the process of building custom databases based on accessions. 

We should move the `add_genomes_to_snpeff_config_file()` call [within snpeff.py](https://github.com/broadinstitute/viral-ngs/blob/03b8368e394f5e058ce575256fdc65bf727c9053/tools/snpeff.py#L104-L109) to be after the `execute('build')` call, and make it contingent on a successful exit code from `snpEff build`.

--- a/tools/snpeff.py
+++ b/tools/snpeff.py
@@ -9,6 +9,7 @@ import os
 import tempfile
 import logging
 import subprocess
+import shutil
 
 # third-party
 import pysam
@@ -41,7 +42,7 @@ class SnpEff(tools.Tool):
         super(SnpEff, self).__init__(install_methods=install_methods)
 
     def version(self):
-        return "4.1"
+        return TOOL_VERSION
 
     def execute(self, command, args, JVMmemory=None, stdin=None, stdout=None, stderr=None):    # pylint: disable=W0221
         if not JVMmemory:
@@ -59,7 +60,7 @@ class SnpEff(tools.Tool):
             ] + args
 
         _log.debug(' '.join(tool_cmd))
-        return util.misc.run_and_print(tool_cmd, stdin=stdin, stderr=stderr, buffered=True, silent=("databases" in command), check=True)
+        return util.misc.run_and_print(tool_cmd, stdin=stdin, stderr=stderr, buffered=True, silent=command in ("databases","build"), check=True)
 
     def has_genome(self, genome):
         if not self.known_dbs:
@@ -101,12 +102,40 @@ class SnpEff(tools.Tool):
                 removeSeparateFiles=False
             )
 
-            add_genomes_to_snpeff_config_file(config_file, [(databaseId, sortedAccessionString, sortedAccessionString)])
+            # create a temp config file in the same location as the original
+            # since the data dir for the built database has a relative
+            # location by default
+            tmp_config_file = config_file+".temp"
+
+            # build a new snpEff database using the downloaded genbank file
+            # Note that if the snpEff command fails for some reason,
+            # we should take care to not include an entry for the database
+            # in the config file, which is the record of installed databases.
+            # In the event the build fails, the config file must reflect
+            # databases present in the data_dir, otherwise calls to execute snpEff will
+            # fail if the data_dir lacks databases for entries listed in the config file.
+
+            # make a temp copy of the old config file
+            shutil.copyfile(config_file, tmp_config_file)
+
+            # add the new genome to the temp config file
+            # since snpEff will not attempt to build unless the
+            # database is in the config file
+            add_genomes_to_snpeff_config_file(tmp_config_file, [(databaseId, sortedAccessionString, sortedAccessionString)])
+
+            try:
+                args = ['-genbank', '-v', databaseId, "-c", tmp_config_file]
+                self.execute('build', args, JVMmemory=JVMmemory)
+            except:
+                # remove temp config file if the database build failed
+                shutil.unlink(tmp_config_file)
+                raise
+
+            # copy the temp config including the built database 
+            # if the execute('build') command did not raise an exception
+            shutil.move(tmp_config_file, config_file)
             self.known_dbs.add(databaseId)
             self.installed_dbs.add(databaseId)
-
-            args = ['-genbank', '-v', databaseId]
-            self.execute('build', args, JVMmemory=JVMmemory)
 
     def available_databases(self):
         # do not capture stderr, since snpEff writes 'Picked up _JAVA_OPTIONS'

